### PR TITLE
Add framework filter to solution list

### PR DESCRIPTION
**Pull request by RahulJha-mastek**
Story [AB#11203](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/11203)
Task [AB#12930](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/12930)
Task [AB#12931](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/12931)

Add framework filter to solution list and update ListAsync() method with new frameworkId parameter

### CODE REVIEW

**Review by bjss-andrew — DISMISSED** [review]

File: src/NHSD.BuyingCatalogue.SolutionLists.Persistence/Repositories/SolutionListRepository.cs [hunk 1]
```diff
@@ -29,10 +30,12 @@ public sealed class SolutionListRepository : ISolutionListRepository
...
                            ON cap.Id = sc.CapabilityId
               LEFT OUTER JOIN dbo.FrameworkSolutions AS fs
-                           ON sol.Id = fs.SolutionId
+                           ON sol.Id = fs.SolutionId 
+                            
              WHERE ps.[Name] = 'Published'
                AND ISNULL(fs.IsFoundation, 0) = COALESCE(@foundationOnly, fs.IsFoundation, 0)
-               AND ci.SupplierId = ISNULL(@supplierId, ci.SupplierId);";
+               AND ci.SupplierId = ISNULL(@supplierId, ci.SupplierId)
+               AND ISNULL(fs.FrameworkId, '') = COALESCE(@frameworkId, fs.FrameworkId, '');";
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.SolutionLists.Persistence/Repositories/SolutionListRepository.cs:38]: `FrameworkId` has a `NOT NULL` constraint so we can omit the `ISNULL` here. On that basis, we can also simplify the `COALESCE` to an `ISNULL` so the condition will be similar to the line above for `SupplierId`, as below:

```sql
AND fs.FrameworkId = ISNULL(@frameworkId, fs.FrameworkId);
```

**Review by bjss-chris — CHANGES_REQUESTED** [review]
Good work @RahulJha-mastek – just one recommendation to simplify the `WHERE` clause of the SQL statement.

File: src/NHSD.BuyingCatalogue.SolutionLists.Persistence/Repositories/SolutionListRepository.cs (hunk 1 above)
Comment (RahulJha-mastek) [reply src/NHSD.BuyingCatalogue.SolutionLists.Persistence/Repositories/SolutionListRepository.cs:38]: In query there is `LEFT outer join` on `dbo.FrameworkSolutions` table, which is making `frameworkId`  column null in where condition. And in SQL `NULL=NULL evaluates to false`, beacuse of which some tests are failing.
That is why I have used `AND ISNULL(fs.FrameworkId, '') = COALESCE(@frameworkId, fs.FrameworkId, '')` same as `IsFoundation` column which is also `NOT NULL`

Comment (RahulJha-mastek) [reply src/NHSD.BuyingCatalogue.SolutionLists.Persistence/Repositories/SolutionListRepository.cs]: Updated Solution query with `INNER JOIN` between `Solution` and `FrameworkSolutions` table

File: src/NHSD.BuyingCatalogue.SolutionLists.Persistence/Repositories/SolutionListRepository.cs [hunk 2]
```diff
@@ -29,13 +30,13 @@ public sealed class SolutionListRepository : ISolutionListRepository
                            ON sol.Id = sc.SolutionId
                    INNER JOIN dbo.Capability AS cap
                            ON cap.Id = sc.CapabilityId
-              LEFT OUTER JOIN dbo.FrameworkSolutions AS fs
+                   INNER JOIN dbo.FrameworkSolutions AS fs
                            ON sol.Id = fs.SolutionId
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.SolutionLists.Persistence/Repositories/SolutionListRepository.cs:35]: Can you remove the blank line here? I don't think we need it.

**Review by bjss-chris — CHANGES_REQUESTED** [review]
Just one tiny request but otherwise looks good. 👍

**Review by bjss-chris — APPROVED** [review]

File: src/NHSD.BuyingCatalogue.SolutionLists.Persistence/Repositories/SolutionListRepository.cs (hunk 2 above)
Comment (RahulJha-mastek) [reply src/NHSD.BuyingCatalogue.SolutionLists.Persistence/Repositories/SolutionListRepository.cs]: Removed Blank line in query

**Pull request merged**